### PR TITLE
Replace increment type with a storage strategy

### PR DESCRIPTION
**Pull request by alcaeus**
This PR deprecates the `increment` type that can be used in ODM, replacing it instead with a more flexible `strategy` option for fields.

First of all, the `increment` type was supposed to handle int and float values at the same time. There was no option to ensure a value read from the database was cast to an int (which could lead to unexpected floats appearing in the application if one wasn't careful).
Furthermore, for collections there is a `strategy` option which controls how the elements in a collection are written to the database. This kind of write control is exactly what should be used for increment fields as well.

At first glance, there are no BC breaks to this. When the `ClassMetadataInfo` class encounters a mapping with type `increment`, it will automatically set the strategy to `increment`, regardless of any other strategy being set. The type is left unchanged to keep the current type cast behavior. Users are encouraged to switch their increment mappings to the new strategy field.

Open tasks:
- [x] Write proper documentation about the new strategy, add deprecation notes for the `increment` type
- [x] Collection strategies still use magic strings, these should be converted to class constants for good measure
- [x] The validation which strategy is acceptable for which mapping should probably be generalized and put in one place to ensure consistency.

@malarzm: I'd appreciate your guidance on the last two tasks since you've spent quite some time with collections ;)

### CODE REVIEW

File: doctrine-mongo-mapping.xsd [hunk 1]
```diff
@@ -64,7 +64,7 @@
     <xs:attribute name="id" type="xs:boolean" default="false" />
     <xs:attribute name="name" type="xs:NMTOKEN" />
     <xs:attribute name="type" type="xs:NMTOKEN" default="string" />
-    <xs:attribute name="strategy" type="xs:NMTOKEN" default="pushAll" />
+    <xs:attribute name="strategy" type="xs:NMTOKEN" default="set" />
```
Comment (malarzm) [inline doctrine-mongo-mapping.xsd]: Why this change?

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/Id.php [hunk 2]
```diff
@@ -18,6 +18,7 @@
  */
 
 namespace Doctrine\ODM\MongoDB\Mapping\Annotations;
+use Doctrine\ODM\MongoDB\Mapping\ClassMetadataInfo;
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Mapping/Annotations/Id.php]: Not needed?

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/Increment.php [hunk 3]
```diff
@@ -26,4 +26,5 @@
 final class Increment extends AbstractField
 {
     public $type = 'increment';
+    public $strategy = 'increment';
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Mapping/Annotations/Increment.php]: Should we add additional note to deprecation message?

File: doctrine-mongo-mapping.xsd (hunk 1 above)
Comment (malarzm) [inline doctrine-mongo-mapping.xsd]: Ah ok, got it after looking further :) Won't it break collections though? (note: I'm n00b when it comes to XMLs and theiir stuff)

**Comment by malarzm** [comment]
Like the approach and where it can go :) I'll elaborate on later today

File: doctrine-mongo-mapping.xsd (hunk 1 above)
Comment (alcaeus) [inline doctrine-mongo-mapping.xsd]: This type only applies to scalar fields, the association types have a different mapping. Technically, this field was not used to begin with ;)

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/Id.php (hunk 2 above)
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Mapping/Annotations/Id.php]: Whoopsie.

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/Increment.php (hunk 3 above)
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Mapping/Annotations/Increment.php]: Yes, I'll add a bit about using increment strategy instead of increment type.

File: lib/Doctrine/ODM/MongoDB/Mapping/MappingException.php [hunk 4]
```diff
@@ -275,4 +275,15 @@ public static function referenceManySortMustNotBeUsedWithNonSetCollectionStrateg
...
         return new self("ReferenceMany's sort can not be used with addToSet and pushAll strategies, $strategy used in $className::$fieldName");
     }
+
+    /**
+     * @param string $className
+     * @param string $fieldName
+     * @param string $strategy
+     * @return MappingException
+     */
+    public static function invalidStorageStrategy($className, $fieldName, $strategy)
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Mapping/MappingException.php]: I'd pass type here as well, optionally we could also throw in allowed strategies and propose correct one basing on Levenshtein distance? But it can be too much :D

**Comment by malarzm** [comment]
> Collection strategies still use magic strings, these should be converted to class constants for good measure

`ClassMetadataInfo::STORAGE_STRATEGY_ATOMIC_SET_ARRAY`? I'd pass on this :) Also it's easier to write `atomicSetArray` in mapping than calling constant I think. Also for internal usage they're pretty much hidden by [CollectionHelper](https://github.com/doctrine/mongodb-odm/blob/453ef86b8c499b05fefe9b67347b8b9c06e3f176/lib/Doctrine/ODM/MongoDB/Utility/CollectionHelper.php)

> The validation which strategy is acceptable for which mapping should probably be generalized and put in one place to ensure consistency.

:+1: on this, end of `ClassMetadataInfo::mapField` seems like a good place for this especially there are no early returns. We could put some `switch` over there that would check certain types if their strategy is allowed (and throw exception if not) and by default unset whatever was passed there.

File: docs/en/reference/storage-strategies.rst [hunk 5]
```diff
@@ -1,29 +1,37 @@
...
+class. The ``increment`` strategy cannot be used for collections.
 
-``addToSet``
-------------
+increment
+---------
+
+The ``increment`` strategy does not apply to collections but can be used for
+``int`` and ``float`` fields. When using the ``increment`` strategy, the field
+value will be updated using an `$inc`_ operator.
```
Comment (jmikola) [inline docs/en/reference/storage-strategies.rst]: I think we can write "the `$inc` operator" here.

File: lib/Doctrine/ODM/MongoDB/Mapping/ClassMetadataInfo.php [hunk 6]
```diff
@@ -1227,6 +1245,55 @@ public function mapField(array $mapping)
     }
 
     /**
+     * Validates the storage strategy of a mapping for consistency
+     * @param array $mapping
+     * @throws \Doctrine\ODM\MongoDB\Mapping\MappingException
+     */
+    private function validateStorageStrategy(array &$mapping)
```
Comment (jmikola) [inline lib/Doctrine/ODM/MongoDB/Mapping/ClassMetadataInfo.php]: "Validate" doesn't convey that we're modifying the `$mapping` argument. Perhaps "apply" would be more appropriate?

File: lib/Doctrine/ODM/MongoDB/Persisters/PersistenceBuilder.php [hunk 7]
```diff
@@ -246,31 +241,28 @@ public function prepareUpsertData($document)
...
                 // If we have a new embedded document then lets set the whole thing
                 if ($new && $this->uow->isScheduledForInsert($new)) {
                     $updateData['$set'][$mapping['name']] = $this->prepareEmbeddedDocumentValue($mapping, $new);
-
-                // If we don't have a new value then do nothing on upsert
-                } elseif ( ! $new) {
-
-                // Update existing embedded document
-                } else {
+                } elseif ( $new) {
```
Comment (jmikola) [inline lib/Doctrine/ODM/MongoDB/Persisters/PersistenceBuilder.php]: Spacing seems a bit odd here.

File: lib/Doctrine/ODM/MongoDB/Mapping/ClassMetadataInfo.php [hunk 8]
```diff
@@ -1227,6 +1245,55 @@ public function mapField(array $mapping)
...
+    {
+        if (! isset($mapping['type']) || isset($mapping['id'])) {
```
Comment (jmikola) [inline lib/Doctrine/ODM/MongoDB/Mapping/ClassMetadataInfo.php]: Should the `!` be surrounded by spaces on both sides for CS?

Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Mapping/ClassMetadataInfo.php]: we lost with PSR after all :(

File: docs/en/reference/storage-strategies.rst [hunk 9]
```diff
@@ -1,29 +1,37 @@
...
 
-Collection Strategies
-=====================
+Storage Strategies
+==================
 
-Doctrine MongoDB ODM implements four different strategies for persisting changes
-to collections of embedded documents or references. These strategies apply to
-the following mapping types:
+Doctrine MongoDB ODM implements seven different strategies for persisting changes
```
Comment (malarzm) [inline docs/en/reference/storage-strategies.rst]: How about changing seven to several? I noticed "four" only lately and haven't got to changing it to "six", I suppose it'll be the same with "seven" when we add eighth one :D

File: lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php [hunk 10]
```diff
@@ -1043,7 +1043,7 @@ private function prepareQueryElement($fieldName, $value = null, $class = null, $
             return array($fieldName, $value);
         }
 
-        if (isset($mapping['strategy']) && CollectionHelper::isHash($mapping['strategy'])
+        if ($mapping['type'] == 'many' && isset($mapping['strategy']) && CollectionHelper::isHash($mapping['strategy'])
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php]: Now that all fields have `strategy` set we could get rid of all `isset($mapping['strategy'])` checks, right? For reference I introduced few with https://github.com/doctrine/mongodb-odm/pull/1177

File: lib/Doctrine/ODM/MongoDB/Utility/CollectionHelper.php [hunk 11]
```diff
@@ -70,6 +71,14 @@ public static function isList($strategy)
      */
     public static function usesSet($strategy)
     {
-        return in_array($strategy, array('set', 'setArray', 'atomicSet', 'atomicSetArray'));
+        return in_array(
+            $strategy,
+            array(
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Utility/CollectionHelper.php]: Short array syntax?

File: tests/Doctrine/ODM/MongoDB/Tests/Mapping/Driver/AbstractDriverTest.php [hunk 12]
```diff
@@ -57,7 +58,8 @@ public function testDriver()
             'isOwningSide' => true,
             'nullable' => false,
             'unique' => true,
-            'sparse' => true
+            'sparse' => true,
+            'strategy' => 'set',
```
Comment (malarzm) [inline tests/Doctrine/ODM/MongoDB/Tests/Mapping/Driver/AbstractDriverTest.php]: This should use constant :)

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/Increment.php (hunk 3 above)
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Mapping/Annotations/Increment.php]: :+1:

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/AbstractField.php [hunk 13]
```diff
@@ -20,6 +20,7 @@
 namespace Doctrine\ODM\MongoDB\Mapping\Annotations;
 
 use Doctrine\Common\Annotations\Annotation;
+use Doctrine\ODM\MongoDB\Mapping\ClassMetadataInfo;
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Mapping/Annotations/AbstractField.php]: We don't need this I think?

**Comment by malarzm** [comment]
Merged manually in https://github.com/doctrine/mongodb-odm/commit/b1cbaf8d674a6fe8d1d6d0db88ce362b8e15b0ec
